enhancements/20250612-volume-online-expansion.md

```diff
@@ -0,0 +1,295 @@
+
+## Proposal
+
+### User Stories
+
+Without this feature, expanding a volume requires shutting down the Virtual Machine or detaching the volume in the guest cluster, both of which result in downtime.
+
+### API Changes
+
+## Design
+
+### Implementation Overview
+
+#### CRD
+
+1. **Enable `ExpandDisks` FeatureGate for `KubeVirt` Resource**
+    ```yaml
+    apiVersion: kubevirt.io/v1
+    kind: KubeVirt
+    metadata:
+      name: kubevirt
+      namespace: harvester-system
+    spec:
+      configuration:
+        additionalGuestMemoryOverheadRatio: "1.5"
+        developerConfiguration:
+          featureGates:
+            - LiveMigration
+            - HotplugVolumes
+            - HostDevices
+            - GPU
+            - CPUManager
+            - VMPersistentState
+            - ExpandDisks
+        network:
+          ...
+        vmStateStorageClass: vmstate-persistence
+    ```
+    - This enables online resize support on the KubeVirt side. See [KubeVirt PR #5981](https://github.com/kubevirt/kubevirt/pull/5981).
+
+2. **Introduce a New Harvester Setting: `csi-online-expand-validation`**
+    ```yaml
+    apiVersion: harvesterhci.io/v1beta1
+    default: '{"driver.longhorn.io":true}'
+    kind: Setting
+    metadata:
+      name: csi-online-expand-validation
+    status: {}
+    value: '{"topolvm.io":true,"hostpath.csi.k8s.io":true}'
+    ```
+    - By default, Longhorn is enabled for volume online expansion, and this setting is immutable:
+        - The Harvester webhook will handle differences between the v1 and v2 Longhorn engines.
+        - The Longhorn v1 engine fully supports volume expansion.
+        - The Longhorn v2 engine does not support volume expansion, regardless of whether the volume is attached or detached.
+    - For third-party storage providers, users can specify whether their CSI driver supports volume online expansion:
+        - By default, Harvester will reject volume online expansion requests for third-party storage.
+        - Harvester will only allow volume online expansion if the storage provider is explicitly configured as validated.
+
+3. **Update the ClusterRole `harvesterhci.io:csi-driver` with New Rules**
+    ```yaml
+    apiVersion: rbac.authorization.k8s.io/v1
+    kind: ClusterRole
+    metadata:
+      name: harvesterhci.io:csi-driver
+    rules:
+      ...
+      - apiGroups:
+          - harvesterhci.io
+        resources:
+          - networkfilesystems
+          - networkfilesystems/status
+          - settings
+        verbs:
+          - '*'
+    ```
+    - In the guest cluster, the Harvester CSI driver will verify the existence of the `csi-online-expand-validation` setting to determine if the PVC supports online expansion.
+
+
+#### PVC Webhook Validator
+
+- PVC Webhook Validator for resizing:
+  - If the volume is not attached to a running VM:
+    - Return `Permitted`
+  - If the KubeVirt `expandDisks` feature gate is not enabled:
```

Comment:
do we really need this check? the kubevirt configuration is being managed by the harvester managedchart and `ExpandDisks` feature gate would have already been enabled as part of the install / upgrade